rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+		eventTracker: eventTracker,
+		samp:         samp,
+		signer:       signer,
+		storage:      storage,
+		transport:    transport,
+	}
+}
+
+// initEventLookupMap generates the SAMUEL event data lookup map from the
+// passed in rootchain configuration
+func initEventData(
+	configEvent *rootchain.ConfigEvent,
+) eventData {
+	return eventData{
+		payloadType:  configEvent.PayloadType,
+		eventABI:     abi.MustNewEvent(configEvent.EventABI),
+		methodABI:    abi.MustNewABI(configEvent.MethodABI).GetMethod(configEvent.MethodName),
+		localAddress: types.StringToAddress(configEvent.LocalAddress),
+	}
+}
+
+// Start starts the SAMUEL module
+func (s *SAMUEL) Start() error {
+	// Start the event loop for the tracker
+	s.startEventLoop()
+
+	// Register the gossip message handler
+	if err := s.registerGossipHandler(); err != nil {
+		return fmt.Errorf("unable to register gossip handler, %w", err)
+	}
+
+	// Fetch the latest event data
+	startBlock, err := s.getStartBlockNumber()
+	if err != nil {
+		return fmt.Errorf("unable to get start block number, %w", err)
+	}
+
+	// Start the Event Tracker
+	if err := s.eventTracker.Start(startBlock); err != nil {
+		return fmt.Errorf("unable to start event tracker, %w", err)
+	}
+
+	return nil
+}
+
+// getStartBlockNumber determines the starting block for the Event Tracker
+func (s *SAMUEL) getStartBlockNumber() (uint64, error) {
+	startBlock := rootchain.LatestRootchainBlockNumber
+
+	data, exists := s.storage.ReadLastProcessedEvent(s.eventData.localAddress.String())
+	if exists && data != "" {
+		// index:blockNumber
+		values := strings.Split(data, ":")
+		if len(values) < 2 {
+			return 0, fmt.Errorf("invalid last processed event in DB: %v", values)
+		}
+
+		blockNumber, err := strconv.ParseUint(values[1], 10, 64)
+		if err != nil {
+			return 0, fmt.Errorf("unable to parse last processed block number in DB: %w", err)
+		}
+
+		startBlock = blockNumber
+	}
+
+	return startBlock, nil
+}
+
+// registerGossipHandler registers a listener for incoming SAM messages
+// from other peers
+func (s *SAMUEL) registerGossipHandler() error {
+	return s.transport.Subscribe(func(sam *proto.SAM) {
+		// Extract the event data
+		eventPayload, err := getEventPayload(sam.Event.Payload, sam.Event.PayloadType)
+		if err != nil {
+			s.logger.Error(
+				fmt.Sprintf("unable to get event payload with hash %s, %v", sam.Hash, err),
+			)
+
+			return
+		}
+
+		// Convert the proto event to a local SAM
+		localSAM := rootchain.SAM{
+			Hash:      types.BytesToHash(sam.Hash),
+			Signature: sam.Signature,
+			Event: rootchain.Event{
+				Index:       sam.Event.Index,
+				BlockNumber: sam.Event.BlockNumber,
+				Payload:     eventPayload,
+			},
+		}
+
+		// Verify that the hash is correct
+		marshalledEvent, err := localSAM.Event.Marshal()
+		if err != nil {
+			s.logger.Error(
+				fmt.Sprintf("unable to marshal event, %v", err),
+			)
+
+			return
+		}
+
+		hash := crypto.Keccak256(marshalledEvent)
```

Comment:
Why not define Hash method in rootchain.Event?

Reply:
What do you mean?

It's always going to be a keccak hash, regardless of the event type

Reply:
To be honest, this part is a bit long to be called from this method. I thought defining method can simplify the flow.

Reply:
Ah, I understand now what you meant. We always do this coupled operation -> marshal + hash, I've gone over the usages and that's the only reason we need the marshal exported, to hash it.

Moved it to a separate helper, it's much cleaner now:

[9217e48](https://github.com/0xPolygon/polygon-edge/pull/717/commits/9217e48af7bd637e4b88db341a436824859a99df)